crates/ollama/src/ollama_completion_provider.rs

```diff
@@ -0,0 +1,786 @@
+use crate::{GenerateOptions, GenerateRequest, generate};
+use anyhow::{Context as AnyhowContext, Result};
+use gpui::{App, Context, Entity, EntityId, Task};
+use http_client::HttpClient;
+use inline_completion::{Direction, EditPredictionProvider, InlineCompletion};
+use language::{Anchor, Buffer, ToOffset};
+use project::Project;
+use std::{path::Path, sync::Arc, time::Duration};
+
+pub const OLLAMA_DEBOUNCE_TIMEOUT: Duration = Duration::from_millis(75);
+
+pub struct OllamaCompletionProvider {
+    http_client: Arc<dyn HttpClient>,
+    api_url: String,
+    model: String,
+    buffer_id: Option<EntityId>,
+    file_extension: Option<String>,
+    current_completion: Option<String>,
+    pending_refresh: Option<Task<Result<()>>>,
+}
+
+impl OllamaCompletionProvider {
+    pub fn new(http_client: Arc<dyn HttpClient>, api_url: String, model: String) -> Self {
+        Self {
+            http_client,
+            api_url,
+            model,
+            buffer_id: None,
+            file_extension: None,
+            current_completion: None,
+            pending_refresh: None,
+        }
+    }
+
+    /// Updates the model used by this provider
+    pub fn update_model(&mut self, new_model: String) {
+        self.model = new_model;
+    }
+
+    fn build_fim_prompt(&self, prefix: &str, suffix: &str) -> String {
+        // Use model-specific FIM patterns
+        let model_lower = self.model.to_lowercase();
+
+        if model_lower.contains("qwen") && model_lower.contains("coder") {
+            // QwenCoder models use pipes
+            format!("<|fim_prefix|>{prefix}<|fim_suffix|>{suffix}<|fim_middle|>")
```

Comment:
I could be wrong, but I'm pretty sure that it's Ollama's job to abstract over this kind of internal model details. Basically each model's template should already define and use the correct tokens. Looks like Ollama provides support for the FIM (fill-in-the-middle) completion in their API via the optional `suffix` parameter, see [the documentation](https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-completion). :slightly_smiling_face:

Reply:
Thanks! I was hoping for predictions to just work initially, and thought I had it working at one point, but then some FIM tokens started showing up in predictions. Hence I started going down this rabbit hole of handling the different cases... And just learned that it's a whole different ballgame with Zeta.

So it'd be a relief not to worry about this - gonna read up on those docs.

Reply:
Ok, removed the FIM token handling and introduced the suffix param. But now I'm getting some chat completion and not just code :P 

need to spend more time grokking this.